### PR TITLE
IPv6 sriov connectivity test

### PR DESCRIPTION
**Pull request by mmirecki**

### CODE REVIEW

File: conformance/ipv6.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+const ipv6NetworkName = "ipv6network"
+const hostnameLabel = "kubernetes.io/hostname"
+
+// Needed for parsing of podinfo
+type Network struct {
+	Interface string
+	Ips       []string
+}
+
+var _ = FDescribe("sriov", func() {
```
Comment (SchSeba) [inline conformance/ipv6.go]: please remove the `F`

File: conformance/ipv6.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+		Expect(err).ToNot(HaveOccurred())
+		Eventually(func() bool {
+			res, err := cluster.SriovStable(operatorNamespace, clients)
+			Expect(err).ToNot(HaveOccurred())
+			return res
+		}, 3*time.Minute, 1*time.Second).Should(Equal(true))
+	})
+
+	Describe("sriov", func() {
+		var _ = Context("IPv6 configured secondary interfaces on pods", func() {
```
Comment (SchSeba) [inline conformance/ipv6.go]: nit: you can remove the `var _` just leave the `Context`

File: conformance/ipv6.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+				Expect(len(ips)).Should(Equal(1))
+				for _, ip := range ips {
+					pingPod(ip, testNode, ipv6NetworkName)
+				}
+			})
+		})
+	})
+})
+
+func createSriovNetwork() {
```
Comment (SchSeba) [inline conformance/ipv6.go]: Can you please move this function to the `utils` folder

File: conformance/ipv6.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	sriovNetwork := &sriovv1.SriovNetwork{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ipv6NetworkName,
+			Namespace: operatorNamespace,
+		},
+		Spec: sriovv1.SriovNetworkSpec{
+			ResourceName:     resourceName,
+			IPAM:             `{"type": "host-local","ranges": [[{"subnet": "3ffe:ffff:0:01ff::/64"}]],"dataDir": "/run/my-orchestrator/container-ipam-state"}`,
```
Comment (SchSeba) [inline conformance/ipv6.go]: after you move the function please make this one a variable for the function

File: conformance/ipv6.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+			NetworkNamespace: testNamespace,
+		}}
+	err := clients.Create(context.Background(), sriovNetwork)
+	Expect(err).ToNot(HaveOccurred())
+}
+
+
+func createSriovPolicy(node string, sriovDevice string, testNode string) {
```
Comment (SchSeba) [inline conformance/ipv6.go]: Can you please move this function to the `utils` folder

File: conformance/ipv6.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+	for _, net := range nets {
+		if net.Interface != sriovInterfaceName {
+			continue
+		}
+		return net.Ips
+	}
+	return nil
+}
+
+func createTestPod(command []string, args []string, node string, sriovNetworkAttachment string) *k8sv1.Pod {
```
Comment (SchSeba) [inline conformance/ipv6.go:96]: there is a function for this in the `utils/pod` folder can you take a look and reuse it?

File: conformance/ipv6.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+	Expect(err).ToNot(HaveOccurred())
+
+	Eventually(func() k8sv1.PodPhase {
+		runningPod, err := clients.Pods(testNamespace).Get(createdPod.Name, metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		return runningPod.Status.Phase
+	}, 3*time.Minute, 1*time.Second).Should(Equal(k8sv1.PodSucceeded))
+}
+
+func testPodDefintion(command []string, args []string, node string, sriovNetworkAttachment string) *k8sv1.Pod {
```
Comment (SchSeba) [inline conformance/ipv6.go]: there is a function for this in the utils/pod folder can you take a look and reuse it?

File: conformance/ipv6.go [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+		runningPod, err := clients.Pods(testNamespace).Get(createdPod.Name, metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		return runningPod.Status.Phase
+	}, 3*time.Minute, 1*time.Second).Should(Equal(k8sv1.PodRunning))
+	pod, err := clients.Pods(testNamespace).Get(createdPod.Name, metav1.GetOptions{})
+	Expect(err).ToNot(HaveOccurred())
+	return pod
+}
+
+func pingPod(ip string, nodeSelector string, sriovNetworkAttachment string) {
```
Comment (SchSeba) [inline conformance/ipv6.go:113]: can you move this function into the `utils/pod` package?

File: conformance/ipv6.go [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+	Eventually(func() int64 {
+		testedNode, err := clients.Nodes().Get(testNode, metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		resNum, _ := testedNode.Status.Capacity["openshift.io/" + resourceName]
+		capacity, _ := resNum.AsInt64()
+		return capacity
+	}, 3*time.Minute, time.Second).Should(Equal(int64(numVfs)))	
+}
+
+func getSriovNicIps(pod *k8sv1.Pod) []string {
```
Comment (SchSeba) [inline conformance/ipv6.go]: can you move this function into the utils/pod package?

I think other tests will like to use this function

**Review by SchSeba — COMMENTED** [review]
Thanks for your contribution!

I have some comments overall looks good!

File: conformance/ipv6.go (hunk 6 above)
Comment (mmirecki) [reply conformance/ipv6.go:96]: Moved testPodDefintion to utils.
Keeping createTestPod, I don't want to move gomega items to utils.

File: conformance/ipv6.go (hunk 8 above)
Comment (mmirecki) [reply conformance/ipv6.go:113]: The function is mainly about Expect's/Eventually's, so keeping some of it here.
Moved the test pod definitinion functions to utils

File: conformance/ipv6.go (hunk 7 above)
Comment (mmirecki) [reply conformance/ipv6.go]: Done

File: conformance/ipv6.go (hunk 9 above)
Comment (mmirecki) [reply conformance/ipv6.go]: Moved to utils.network

File: conformance/ipv6.go (hunk 5 above)
Comment (mmirecki) [reply conformance/ipv6.go]: Moved to utils.network

File: conformance/ipv6.go (hunk 4 above)
Comment (mmirecki) [reply conformance/ipv6.go]: Done

File: conformance/ipv6.go (hunk 3 above)
Comment (mmirecki) [reply conformance/ipv6.go]: Moved to utils.network

File: conformance/ipv6.go (hunk 2 above)
Comment (mmirecki) [reply conformance/ipv6.go]: Done

File: conformance/ipv6.go (hunk 1 above)
Comment (mmirecki) [reply conformance/ipv6.go]: Done

File: conformance/ipv6.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+	k8sv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const sriovInterfaceName = "net1"
+const numVfs = 5
+const resourceName = "sriovnic"
+const ipv6NetworkName = "ipv6network"
+
+var _ = Describe("sriov", func() {
```
Comment (SchSeba) [inline conformance/ipv6.go]: nit: can you update the describe this all project is for sriov tests

File: conformance/ipv6.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+		err := namespaces.Clean(operatorNamespace, namespaces.Test, clients)
+		Expect(err).ToNot(HaveOccurred())
+		Eventually(func() bool {
+			res, err := cluster.SriovStable(operatorNamespace, clients)
+			Expect(err).ToNot(HaveOccurred())
+			return res
+		}, 3*time.Minute, 1*time.Second).Should(Equal(true))
+	})
+
+	Describe("sriov", func() {
```
Comment (SchSeba) [inline conformance/ipv6.go]: nit: can you update the describe this all project is for sriov tests

File: conformance/ipv6.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	})
+
+	BeforeEach(func() {
+		err := namespaces.Clean(operatorNamespace, namespaces.Test, clients)
+		Expect(err).ToNot(HaveOccurred())
+		Eventually(func() bool {
+			res, err := cluster.SriovStable(operatorNamespace, clients)
+			Expect(err).ToNot(HaveOccurred())
+			return res
+		}, 3*time.Minute, 1*time.Second).Should(Equal(true))
```
Comment (SchSeba) [inline conformance/ipv6.go]: please update the time here to 10 min to cover a possible node reboot if is a Mellanox card

File: conformance/ipv6.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+})
+
+func createSriovPolicy(sriovDevice string, testNode string) {
+	err := network.CreateSriovPolicy(clients, "test-policy-", operatorNamespace, sriovDevice, testNode, numVfs, resourceName)
+	Expect(err).ToNot(HaveOccurred())
+	Eventually(func() bool {
+		stable, err := cluster.SriovStable(operatorNamespace, clients)
+		Expect(err).ToNot(HaveOccurred())
+		return stable
+	}, 5*time.Minute, 1*time.Second).Should(Equal(true))
```
Comment (SchSeba) [inline conformance/ipv6.go]: update the time here to 10 min please

File: pkg/util/network/network.go [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Needed for parsing of podinfo
+type Network struct {
+	Interface string
+	Ips       []string
+}
+
+func CreateSriovNetwork(clientSet *testclient.ClientSet, name string, namespace string, operatorNamespace string, resourceName string, ipam string) error {
```
Comment (SchSeba) [inline pkg/util/network/network.go:19]: The ipam string is not in use here?

File: conformance/ipv6.go [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+	Describe("sriov", func() {
+		Context("IPv6 configured secondary interfaces on pods", func() {
+			It("should be able to ping each other", func() {
+				sriovDevice, err := sriovInfos.FindOneSriovDevice(testNode)
+				Expect(err).ToNot(HaveOccurred())
+				createSriovPolicy(sriovDevice.Name, testNode)
+
+				ipam := `{"type": "host-local","ranges": [[{"subnet": "3ffe:ffff:0:01ff::/64"}]],"dataDir": "/run/my-orchestrator/container-ipam-state"}`
+				err = network.CreateSriovNetwork(clients, ipv6NetworkName, namespaces.Test, operatorNamespace, resourceName, ipam)
+				Expect(err).ToNot(HaveOccurred())
```
Comment (SchSeba) [inline conformance/ipv6.go:58]: Can you add Eventually here to validate the net-attach-def exist?

something like this:
```golang
Eventually(func() error {
netAttDef := &netattdefv1.NetworkAttachmentDefinition{}
return clients.Get(context.Background(), runtimeclient.ObjectKey{Name: mtuNetwork.Name, Namespace: namespaces.Test}, netAttDef)
}, 10*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
```

**Review by SchSeba — CHANGES_REQUESTED** [review]
overall looks good just a few comments and this PR should be ready to go

File: conformance/ipv6.go (hunk 10 above)
Comment (mmirecki) [reply conformance/ipv6.go]: changed to ipv6

File: conformance/ipv6.go (hunk 11 above)
Comment (mmirecki) [reply conformance/ipv6.go]: changed to ipv6

**Review by SchSeba — APPROVED** [review]
/lgtm 

/hold 
@fedepaol can you take a look from your side also?

File: pkg/util/pod/pod.go [hunk 16]
```diff
@@ -43,6 +45,27 @@ func DefineWithHostNetwork() *corev1.Pod {
 	return podObject
 }
 
+// RedefineWithNodeSelector - Update pod definition with node selector
```
Comment (fedepaol) [inline pkg/util/pod/pod.go]: Can you rephrase the comment in a more imperative way? 
Like `RedefineWithNodeSelector updates the pod definition with..". 
Same as per comments below

**Review by fedepaol — COMMENTED** [review]
Just one small nit about idiomatic comments

File: pkg/util/pod/pod.go (hunk 16 above)
Comment (mmirecki) [reply pkg/util/pod/pod.go]: Done
